src/lib/log.c

```diff
@@ -48,3 +103,111 @@ void log_put(uint8_t c)
 		++buf_idx;
 	}
 }
+
+
+/* Send a string to the logger: wait for full lines before actually sending. */
+static void log_puts(const char *s)
+{
+	for (; *s; ++s) {
+		log_put(*s);
+	}
+}
+
+
+/* Send a string to the logger: wait for full lines before actually sending. */
+static int log_vprintf(const char *fmt, va_list args)
+{
+	static char buf[4096];
+
+	int res = vsnprintf(buf, sizeof(buf), fmt, args);
+
+	log_puts(buf);
```

Comment:
Do I understand correctly that the log buffer will now be a mixture of the guest console output and any diagnostics emitted by Hyperkit itself using `fprintf`? I am very much not keen on that arrangement.

Logs from Hyperkit itself should either (depending on the configuration) go to stdio or to ASL directly, never via the log buffering stuff used for the guest console.

It looks like the direct uses of `log_info` and friends are doing the right thing here, it's just the imposition on the `fprintf(stderr, ...` calls which looks wrong.

Reply:
I've made it the other way round: the console lines are assembled until the end of line, and then we send it to the logger.  And the logger sends them to stderr or asl, depending on the command line choice.